### DISCUSSION TITLE
OOM with only a middlin' data set -- how to subdivide?

### DISCUSSION THREAD

**Discussion by vputz** [jax-ml/jax#10672] [Q&A]
OK, I feel dumb.  I need to iterate over a bunch of images for training and testing, like everyone else, but I run out of GPU memory _just creating the training batches_ because allocating a bunch of DeviceArrays to hold the whole data set fills up memory since the data set is larger than GPU memory

(code like
```
batches = [ jnp.array(numpyarray[i*batch_size:(i+1)*batch_size]) for i in range(num_batches)]
```
so truly nothing cosmic; the "numpyarray" is just a giant numpy array in CPU memory and I'm trying to slice it into sections that will become DeviceArrays)

I guess I thought once that was done things would be shuffled in and out of GPU memory as necessary since I don't see any way to free up allocations, but you have to put things into device memory at some point.  I know I'm missing something *really* obvious: how do you use jax to iterate over a data set larger than the GPU?

**Comment by vputz** [comment]
(I've tried leaving everything in numpy and in my iteration do
```
for train_batch in train_batches:
  state, metrics = train(..., jnp.asarray(train_batch[0]), jnp.asarray(train_batch[1]))
```

which sounds like it should work and prevents oom, but is barely faster than the CPU and sometimes even not that.  It's most perplexing. Increasing the batch size actually increased the time; none of this makes sense!

**Reply by YouJiacheng** [reply]
What GPU do you use? And what is the ratio of FLOPs(of model) to data transfer?

**Reply by vputz** [reply]
GPU is a 3070 (consumer) for my testing (they've run on an A100 so we have a variety).  So nominally 20 teraflops and we're on a two-year-old motherboard so while I can't quote data transfer rates, this still seems strange to me that there' SO MUCH of a speed differential with paging as above.

More directly, how _does_ one handle "data sets larger than the device memory"--is it the above, and Jax is meant to shuffle them in and out of device memory at will?  A good code example I can build off of and play with pprof to understand better what's happening or should happen?

**Reply by YouJiacheng** [reply]
By "the ratio of FLOPs(of model) to data transfer", I mean: for 1 byte input(or 1 float point number input), how many float point operations are performed in the model?
If the model only perform a few computation, then the data transfer overhead can be dominant.
For such case, prefetch can be used.
